## PHẦN A — ĐẶT CÂU HỎI

## 10 Câu Hỏi Khám Phá Theo Trực Giác

1. **Có bao nhiêu hành khách trong Titanic?**  
   Tôi dự đoán kết quả sẽ là: có khoảng hơn 800 hành khách.

2. **Số lượng nam hành khách có nhiều hơn nữ hành khách?**  
   Tôi dự đoán kết quả sẽ là: có, nam hành khách có nhiều hơn nữ hành khách.

3. **Phụ nữ có tỷ lệ sống sót cao hơn nam giới không?**  
   Tôi dự đoán kết quả sẽ là: phụ nữ có tỷ lệ sống sót cao hơn rất nhiều so với nam giới.

4. **Hành khách là trẻ con có khả năng sống cao không?**  
   Tôi dự đoán kết quả sẽ là: có, trẻ em (đặc biệt dưới 10 tuổi) sẽ có tỷ lệ sống sót cao.

5. **Những hành khách đi một mình có tỷ lệ sống sót cao không?**  
   Tôi dự đoán kết quả sẽ là: không, người đi một mình sẽ có tỷ lệ sống sót thấp hơn người đi cùng gia đình.

6. **Có bao nhiêu hành khách sống sót?**  
   Tôi dự đoán kết quả sẽ là: có khoảng 500 hành khách sống sót.

7. **Hành khách mua vé tiền cao có được ưu tiên giải cứu không?**  
   Tôi dự đoán kết quả sẽ là: có, người mua vé càng đắt sẽ có tỷ lệ sống sót càng cao.

8. **Giá vé trung bình của khoang hạng 1 là bao nhiêu?**  
   Tôi dự đoán kết quả sẽ là: giá vé trung bình ở khoảng trên 60 USD.

9. **Nam giới lớn tuổi (trên 60) có tỷ lệ sống sót cao không?**  
   Tôi dự đoán kết quả sẽ là: không, nam giới trên 60 hầu như không sống sót.

10. **Tổng số người trong gia đình ảnh hưởng như thế nào đến việc sống sót?**  
    Tôi dự đoán kết quả sẽ là: có, gia đình có 2-4 người có tỷ lệ sống sót cao hơn đi một mình hay gia đình quá lớn.

## PHẦN B — KHÁM PHÁ CẤU TRÚC DỮ LIỆU

In [2]:
import seaborn as sns
import pandas as pd

# Load dataset
titanic = sns.load_dataset('titanic')

titanic.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


### head()
5 dòng đầu tiên cho thấy dữ liệu bao gồm thông tin về hành khách: survived, pclass, sex, age, fare, embarked...

Dữ liệu có cả biến số (numeric) và biến phân loại (categorical). 
Cột `deck` và `age` có thể chứa missing values.

In [3]:
titanic.tail()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
886,0,2,male,27.0,0,0,13.00,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.00,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.45,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.00,C,First,man,True,C,Cherbourg,yes,True
890,0,3,male,32.0,0,0,7.75,Q,Third,man,True,NaN,Queenstown,no,True


### tail()
5 dòng cuối cùng cho thấy cấu trúc dữ liệu nhất quán. 
Một số hành khách có thông tin age và deck bị thiếu.

In [4]:
print("Shape:", titanic.shape)
titanic.info()

Shape: (891, 15)
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    str     
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    str     
 8   class        891 non-null    category
 9   who          891 non-null    str     
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    str     
 13  alive        891 non-null    str     
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), str(5)
memory usage: 80.7 KB


### shape và info()
Dataset có **891 hàng** (hành khách) và **15 cột**.

Có 2 kiểu dữ liệu chính: số (int64, float64) và phân loại (object, category, bool). 
Cột `age`, `deck`, `embarked` có missing values.

In [5]:
titanic.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### describe()
- `survived`: Tỷ lệ sống sót trung bình khoảng 38.4%.
- `age`: Tuổi trung bình ~29.7, cao nhất 80.
- `fare`: Giá vé trung bình ~32.2, nhưng có sự chênh lệch rất lớn (max 512.33).

In [7]:
# Missing values
missing = titanic.isnull().sum()
missing_percent = (titanic.isnull().sum() / len(titanic) * 100).round(2)

missing_df = pd.DataFrame({
    'Số lượng missing': missing,
    'Tỷ lệ (%)': missing_percent
})
missing_df[missing_df['Số lượng missing'] > 0]

,Số lượng missing,Tỷ lệ (%)
age,177,19.87
embarked,2,0.22
deck,688,77.22
embark_town,2,0.22


### Phân tích Missing Values

- **deck**: 688 missing (77.2%) → Thiếu rất nhiều, có thể bỏ cột này.
- **age**: 177 missing (19.9%) → Tương đối nhiều, cần xử lý (impute median hoặc nhóm theo pclass/sex).
- **embarked**: 2 missing (0.22%) → Rất ít, có thể dropna hoặc điền mode.
- **embark_town**: 2 missing (giống embarked).

**Cột có vấn đề lớn nhất:** `deck` (thiếu quá nhiều nên khó sử dụng).

In [8]:
print("Sex:\n", titanic['sex'].value_counts())
print("\nPclass:\n", titanic['pclass'].value_counts())
print("\nEmbarked:\n", titanic['embarked'].value_counts(dropna=False))

Sex:
 sex
male      577
female    314
Name: count, dtype: int64

Pclass:
 pclass
3    491
1    216
2    184
Name: count, dtype: int64

Embarked:
 embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


### Phân tích 3 cột Categorical

- **sex**: Nam (577) chiếm đa số so với Nữ (314) → Tỷ lệ nam/nữ ≈ 65/35.
- **pclass**: Hạng 3 (491) chiếm nhiều nhất, sau đó là hạng 1 (216) và hạng 2 (184). Hành khách nghèo chiếm ưu thế.
- **embarked**: S (644) nhiều nhất (Southampton), C (168), Q (77). Có 2 giá trị NaN.

In [9]:
print("AGE:")
print("Mean   :", titanic['age'].mean().round(2))
print("Median :", titanic['age'].median())
print("Min    :", titanic['age'].min())
print("Max    :", titanic['age'].max())

print("\nFARE:")
print("Mean   :", titanic['fare'].mean().round(2))
print("Median :", titanic['fare'].median())
print("Min    :", titanic['fare'].min())
print("Max    :", titanic['fare'].max())

print("\nSIBSP:")
print("Mean   :", titanic['sibsp'].mean().round(2))
print("Median :", titanic['sibsp'].median())
print("Min    :", titanic['sibsp'].min())
print("Max    :", titanic['sibsp'].max())

AGE:
Mean   : 29.7
Median : 28.0
Min    : 0.42
Max    : 80.0

FARE:
Mean   : 32.2
Median : 14.4542
Min    : 0.0
Max    : 512.3292

SIBSP:
Mean   : 0.52
Median : 0.0
Min    : 0
Max    : 8


### Phân tích 3 cột Numeric

- **age**: Mean ≈ 29.7, Median = 28. Độ tuổi phân bố khá rộng, có trẻ sơ sinh (0.42) và người già (80).
- **fare**: Mean ≈ 32.2, Median = 14.45. Phân bố lệch phải mạnh (max 512), cho thấy một số vé rất đắt.
- **sibsp**: Mean ≈ 0.54, Median = 0. Hầu hết hành khách đi một mình (không có anh em/vợ chồng).

**Điểm bất thường**: Giá vé có outlier rất lớn (512), fare phân bố không đồng đều.

## PHẦN C — PHÂN TÍCH CƠ BẢN

In [10]:
import seaborn as sns
import pandas as pd

titanic = sns.load_dataset('titanic')

# Tỷ lệ sống sót toàn tàu
survival_rate = titanic['survived'].mean() * 100
total_survived = titanic['survived'].sum()
total_passengers = len(titanic)

print(f"Tổng số hành khách: {total_passengers}")
print(f"Số người sống sót: {total_survived}")
print(f"Tỷ lệ sống sót toàn tàu: {survival_rate:.2f}%")

Tổng số hành khách: 891
Số người sống sót: 342
Tỷ lệ sống sót toàn tàu: 38.38%


### Tỷ lệ sống sót toàn tàu

**Kết quả:** Trong tổng số **891** hành khách, chỉ có khoảng **342** người sống sót, tương đương **38.38%**.

**Giải thích:**  
Con số này cho thấy thảm họa Titanic rất nghiêm trọng — hơn **61%** hành khách đã thiệt mạng. Đây là lý do tại sao "phụ nữ và trẻ em trước" là nguyên tắc nổi tiếng trong sự kiện này.

In [11]:
# Tỷ lệ sống sót theo giới tính
survival_by_sex = titanic.groupby('sex')['survived'].mean() * 100
print(survival_by_sex)

sex
female    74.203822
male      18.890815
Name: survived, dtype: float64


### Tỷ lệ sống sót theo Giới tính

- **Nữ**: 74.04% sống sót  
- **Nam**: 18.89% sống sót  

**Nhận xét:** Phụ nữ có tỷ lệ sống sót cao gấp gần **4 lần** nam giới. Điều này phản ánh nguyên tắc "phụ nữ và trẻ em trước" đã được thực thi trên tàu Titanic.

In [12]:
# Tỷ lệ sống sót theo hạng vé
survival_by_pclass = titanic.groupby('pclass')['survived'].mean() * 100
print(survival_by_pclass)

pclass
1    62.962963
2    47.282609
3    24.236253
Name: survived, dtype: float64


### Tỷ lệ sống sót theo Hạng vé (Pclass)

- **Hạng 1**: 62.96%  
- **Hạng 2**: 47.28%  
- **Hạng 3**: 24.24%  

**Nhận xét:** Hành khách càng mua vé đắt (hạng cao) thì càng có tỷ lệ sống sót cao hơn rõ rệt. Hạng 1 có tỷ lệ sống sót gấp **2.6 lần** so với hạng 3.

In [13]:
# Câu 1: Có bao nhiêu hành khách?
print("Tổng số hành khách:", len(titanic))

# Câu 3: Phụ nữ có tỷ lệ sống sót cao hơn nam không?
print("\nTỷ lệ sống sót theo giới tính:")
print(titanic.groupby('sex')['survived'].mean() * 100)

# Câu 7: Vé càng đắt càng dễ sống sót?
print("\nTỷ lệ sống sót theo hạng vé:")
print(titanic.groupby('pclass')['survived'].mean() * 100)

Tổng số hành khách: 891

Tỷ lệ sống sót theo giới tính:
sex
female    74.203822
male      18.890815
Name: survived, dtype: float64

Tỷ lệ sống sót theo hạng vé:
pclass
1    62.962963
2    47.282609
3    24.236253
Name: survived, dtype: float64


### Trả lời 3 câu hỏi từ Phần A

**Câu 1:** Có bao nhiêu hành khách trong Titanic?  
→ **Kết quả:** 891 hành khách.

**Câu 3:** Phụ nữ có tỷ lệ sống sót cao hơn nam giới không?  
→ **Kết quả:** Đúng. Phụ nữ: **74.0%**, Nam: **18.9%**.

**Câu 7:** Hành khách mua vé tiền cao có được ưu tiên không?  
→ **Kết quả:** Đúng. Hạng 1 (vé đắt): **62.96%**, Hạng 3 (vé rẻ): chỉ **24.24%**.

## Đối chiếu Dự đoán ban đầu vs Thực tế

| STT | Câu hỏi | Dự đoán | Thực tế | Đánh giá |
|-----|--------|--------|--------|---------|
| 1   | Tổng số hành khách | >800 | 891 | **Đúng** |
| 3   | Phụ nữ sống sót cao hơn nam | Cao hơn rất nhiều | 74% vs 19% | **Đúng** |
| 7   | Vé đắt → sống sót cao | Có | Hạng 1: 63% | **Đúng** |
| ... | ... | ... | ... | ... |

**Nhận xét tổng quát:**  
Hầu hết các dự đoán ban đầu đều **đúng** hoặc rất gần với thực tế. Điều này cho thấy "phụ nữ và trẻ em được ưu tiên" cùng "người giàu có lợi thế" là khá chính xác. Tuy nhiên, vẫn cần phân tích sâu hơn (ví dụ: vai trò của tuổi tác và deck) để hiểu rõ hơn về sự kiện lịch sử này.

## PHẦN D — CÂU HỎI MỚI (SAU PHÂN TÍCH)

1. **Trẻ em (dưới 12 tuổi) có tỷ lệ sống sót cao hơn người lớn không?**  

2. **Trong cùng một hạng vé, phụ nữ vẫn có lợi thế sống sót hơn nam giới nhiều không?**  

3. **Hành khách lên tàu tại cảng Cherbourg (C) có tỷ lệ sống sót cao hơn so với Southampton (S) và Queenstown (Q) không?**  

4. **Người có cabin (deck) được ghi nhận có tỷ lệ sống sót cao hơn những người không có thông tin deck không?**  

5. **Giá vé (Fare) có tương quan mạnh với khả năng sống sót không? (không chỉ nhìn hạng vé)**